### PR TITLE
Fix tag handling for multiple tag lists

### LINKED ISSUE #3183: Can't give multiple --tag switches
Hi again :wave:

Since I’m using master as close as possible, [one of the recent commits](https://github.com/restic/restic/commit/ae441d31348ab714fbe3e0a09d2bc68847617dec) turned out to be backwards-incompatible.

Now if we give `--tag tag1 --tag tag2` while getting a backup, restic ignores previous switches and only applies last given tag.

Help switch after the change says:

```
add tags for the new snapshot in the format `tag[,tag,...]` (can be specified multiple times) (default [])
```

for --tag switch, which is a bit confusing. I can only specify multiple tags, not the --tag switch multiple times.

(Redirected from [forum](https://forum.restic.net/t/a-small-question-about-new-tag-syntax/3384))

### PR DESCRIPTION
**Pull request by SkYNewZ**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Specifying the `--tag' option several times now works as expected.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Fix https://github.com/restic/restic/issues/3183.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by SkYNewZ** [comment]
Seems good for me now.

```
restic via 🐹 v1.15.5 using ☁️  default/groups-terraform-sync-dtep on  fix/3183 
➜ go build ./cmd/restic

restic via 🐹 v1.15.5 using ☁️  default/groups-terraform-sync-dtep on  fix/3183 
➜ ./restic version
restic 0.11.0-dev (compiled manually) compiled with go1.15.5 on linux/amd64

restic via 🐹 v1.15.5 using ☁️  default/groups-terraform-sync-dtep on  fix/3183 
➜ ./restic backup --tag foo --tag bar --tag hello --hostname me /home/quentin/Downloads/test-restic
Flag --hostname has been deprecated, use --host
repository 1a5af6f7 opened successfully, password is correct
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            4 new,     0 changed,     0 unmodified
Added to the repo: 1.865 KiB

processed 1 files, 6 B in 0:04
snapshot 806e08ea saved

restic via 🐹 v1.15.5 using ☁️  default/groups-terraform-sync-dtep on  fix/3183 
➜ ./restic snapshots                                                                               
repository 1a5af6f7 opened successfully, password is correct
ID        Time                 Host                 Tags           Paths
------------------------------------------------------------------------------------------------------------
e4e6d726  2020-11-03 19:30:14  MBP-de-Quentin.Home                 /Users/quentin/Downloads/ghost_backup.sql
20791c35  2020-12-26 01:00:08  kubernetes           feedly         /feedly-export.opml
b27e76ee  2020-12-27 01:00:08  kubernetes           feedly         /feedly-export.opml
383365ba  2020-12-28 01:00:08  kubernetes           feedly         /feedly-export.opml
806e08ea  2020-12-28 16:36:18  me                   foo,bar,hello  /home/quentin/Downloads/test-restic
------------------------------------------------------------------------------------------------------------
5 snapshots

restic via 🐹 v1.15.5 using ☁️  default/groups-terraform-sync-dtep on  fix/3183 
➜ ./restic cat snapshot 806e08ea
repository 1a5af6f7 opened successfully, password is correct
{
  "time": "2020-12-28T16:36:18.500153975+01:00",
  "tree": "7cc00eddade4fb69ae54b2fc6a9a48b6c97a3a077c3b948a958fb1bd28d1917f",
  "paths": [
    "/home/quentin/Downloads/test-restic"
  ],
  "hostname": "me",
  "username": "quentin",
  "uid": 1000,
  "gid": 1000,
  "tags": [
    "foo",
    "bar",
    "hello"
  ]
}
```

File: changelog/unreleased/issue-3183 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Bugfix: Allow specify --tag option several times and the resulting backup must contain all the specified tags.
```
Comment (aawsome) [inline changelog/unreleased/issue-3183:1]: Suggestion: "Bugfix: Allow specify --tag option several times for backup command"

File: internal/archiver/archiver.go [hunk 2]
```diff
@@ -717,7 +717,7 @@ func resolveRelativeTargets(filesys fs.FS, targets []string) ([]string, error) {
 
 // SnapshotOptions collect attributes for a new snapshot.
 type SnapshotOptions struct {
-	Tags           restic.TagList
+	Tags           restic.TagLists
```
Comment (aawsome) [inline internal/archiver/archiver.go:720]: I would keep the `TagList` in the archiver code and would use the `Flatten()` method in `cmd_backup.go`, line 685.

File: internal/restic/tag_list_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+					"tag3",
+					"tag4",
+				},
+			},
+			want: &TagList{"tag1", "tag2", "tag3", "tag4"},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.l.Flatten(); !reflect.DeepEqual(got, tt.want) {
```
Comment (aawsome) [inline internal/restic/tag_list_test.go:31]: maybe use `test.Equals` from `internal/test`?

File: internal/restic/tag_list.go [hunk 4]
```diff
@@ -40,6 +40,23 @@ func (l TagLists) String() string {
 	return fmt.Sprintf("%v", []TagList(l))
 }
 
+// Flatten returns the list of all tags provided in TagLists
+func (l TagLists) Flatten() *TagList {
```
Comment (MichaelEischer) [inline internal/restic/tag_list.go:44]: Please return the `TagList` as value and not a pointer. This also allows removing the `len(l)==0` special case.

File: internal/restic/tag_list_test.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	"testing"
+)
+
+func TestTagLists_Flatten(t *testing.T) {
+	tests := []struct {
+		name string
+		l    TagLists
+		want *TagList
+	}{
+		{
```
Comment (MichaelEischer) [inline internal/restic/tag_list_test.go:14]: Please add a test case which tries to flatten an empty `TagLists`

File: changelog/unreleased/issue-3183 (hunk 1 above)
Comment (MichaelEischer) [inline changelog/unreleased/issue-3183:1]: Please remove this changelog file and add a link to this PR to the changelog `issue-2688`. The bug hasn't been part of a released restic version. There's no need to inform regular users about bugs that were introduced and fixed between releases.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
The set, add and remove options of the `tag` command are also affected by the same problem. Please also change these to `TagLists`.

File: cmd/restic/integration_test.go [hunk 6]
```diff
@@ -662,7 +662,7 @@ func TestBackupTags(t *testing.T) {
 		"expected no tags, got %v", newest.Tags)
 	parent := newest
 
-	opts.Tags = []string{"NL"}
+	opts.Tags = []restic.TagList{[]string{"NL"}}
```
Comment (aawsome) [inline cmd/restic/integration_test.go:665]: `opts.Tags = restic.TagLists{[]string{"NL"}}` would be a bit more precise, but ok :wink:

**Review by aawsome — APPROVED** [review]
LGTM!

File: cmd/restic/integration_test.go (hunk 6 above)
Comment (SkYNewZ) [reply cmd/restic/integration_test.go]: Fixed 😉

**Comment by SkYNewZ** [comment]
> LGTM!

Thanks !

**Review by MichaelEischer — APPROVED** [review]
LGTM.

> Thanks ! Sorry for introducing this bug.

No problem at all. It didn't catch it either during review.
